### ISSUE TITLE
BUG: Error when leaving route from function inside isReady condition

### ISSUE THREAD

**Issue by miri-am** [meteorhacks/flow-router#103]
I am not sure if this is related to routing or a more general question. When I have a route like the following that subscribes to a published collection.

```
FlowRouter.route('/settings', {
  subscriptions: function(params) {
      this.register('institutions', Meteor.subscribe('institutions'));
  },
  action: function(params) {
      FlowLayout.render("appBody", {
          area: "settings"
      });
  }
});
```

And my template and helpers look like that:

settings.html
```
<template name="settings">
    {{#if isReady}}
        <p>{{getInstitutionName}}</p>
    {{/if}}
</template>
```

settings.js
```
Template.settings.helpers({
	getInstitutionName: function(){
		if(Meteor.userId()){
			return Institutions.findOne(Meteor.user().profile.institution_id).name;
		}
	},
    isReady: function(sub) {
      if(sub) {
        return FlowRouter.subsReady(sub);
      } else {
        return FlowRouter.subsReady();
      }
    }	
})
```
Then I receive the following error when **leaving** the route.

```
Exception in template helper: TypeError: Cannot read property 'name' of undefined
    at Object.Template.settings.helpers.getInstitutionName (http://localhost:3000/client/templates/settings/settings.js?60d79c554ecab2e8a6522c2526b261e120a1405f:9:69)
```

I can resolve this by adjusting my helper like in the following and ensuring that the findOne() returns something. I just wondered if this is the correct way or if my `isReady` helper should avoid that getInstitutionName is called again?

```
getInstitutionName: function(){
    if(Meteor.userId() && Institutions.findOne(Meteor.user().profile.institution_id)){
	return Institutions.findOne(Meteor.user().profile.institution_id).name;
    }
},

**Comment by wI2L** [comment]
The `isReady` helper will ensure that the rest won't render until the subscriptions is ready.
You coud also omit this isReady helper, and the institution name will render as soon as the data is recaived from the publication, reactively.

However this is not related to flow-router.

You could do something like this using `with`


**Helper**
```
institution: function(){
    if (Meteor.user() && Meteor.user().profile){
        return Institutions.findOne(Meteor.user().profile.institution_id);
    }
}
```

**HTML**

```
<template name="settings">
    {{#if isReady}}
       {{#with institution}}
          <p>{{name}}</p>
        {{/with}}
    {{/if}}
</template>
```

**Comment by miri-am** [comment]
With changing routes I mean for example calling `FlowRouter.go("/otherRoute");` I think the problem is that the helpers are rerun when I change routes (because the other route does not subscribe to Institutions). Using "guarding" avoids the error - just thought that there may be other ways to avoid reruns when leaving the route.

**Comment by arunoda** [comment]
Yes. This is a bug and related to #95

**Comment by arunoda** [comment]
Can you check the `fix-reactive-bugs` branch. I did a fix for this.

**Comment by miri-am** [comment]
with this branch some of my routes are not working anymore: 
```There is no route for the path: /changePassword```

**Comment by arunoda** [comment]
Hmm.
Can I have the route definition or a sample app?

**Comment by miri-am** [comment]
sure. It works when using the master branch and does not work when using the fix-reactive-bugs branch
```
FlowRouter.route('/changePassword', {
  subscriptions: function(params) {
  },
  action: function(params) {
      FlowLayout.render("appBody", {
          area: "changePassword"
      });
  }
});
```

**Comment by arunoda** [comment]
I will check this.

**Comment by arunoda** [comment]
Released version `1.8.0` with fixes for this bug.
